pkg/controllers/user/nsserviceaccount/nssvcaccnt.go

```diff
@@ -0,0 +1,82 @@
+package nsserviceaccount
+
+import (
+	"context"
+	"strings"
+
+	"github.com/rancher/rancher/pkg/settings"
+	rv1 "github.com/rancher/types/apis/core/v1"
+	"github.com/rancher/types/config"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+type namespaceSvcAccountHandler struct {
+	serviceAccountLister rv1.ServiceAccountLister
+	serviceAccounts      rv1.ServiceAccountInterface
+}
+
+func Register(ctx context.Context, cluster *config.UserContext) {
+	logrus.Debugf("Registering namespaceSvcAccountHandler for checking default serviceaccount")
+	nsh := &namespaceSvcAccountHandler{
+		serviceAccountLister: cluster.Core.ServiceAccounts("").Controller().Lister(),
+		serviceAccounts:      cluster.Core.ServiceAccounts(""),
+	}
+	cluster.Core.Namespaces("").AddHandler(ctx, "namespaceSvcAccountHandler", nsh.Sync)
+}
+
+func (nsh *namespaceSvcAccountHandler) Sync(key string, ns *corev1.Namespace) (runtime.Object, error) {
+	if ns == nil || ns.DeletionTimestamp != nil {
+		return nil, nil
+	}
+	logrus.Debugf("namespaceSvcAccountHandler: Sync namespace: key=%v", key)
+
+	//handle default svcAccount of system namespaces
+	if err := nsh.handleSystemNS(key); err != nil {
+		logrus.Errorf("namespaceSvcAccountHandler: Sync: error handling default ServiceAccount of namespace key=%v, err=%v", key, err)
+	}
+	return nil, nil
+}
+
+func (nsh *namespaceSvcAccountHandler) handleSystemNS(namespace string) error {
+	if namespace == "kube-system" || namespace == "default" || !nsh.isSystemNS(namespace) {
+		return nil
+	}
+	defSvcAccnt, err := nsh.serviceAccountLister.Get(namespace, "default")
+	if err != nil {
```

Comment:
We aren't checking for NotFound error here, would there be some system-namespaces where the default SA won't exist ?

Reply:
Even if there is any such system ns where the default SA does not exist, there is no need to error out - we will just skip that ns. The main function of the controller is to update AutomountServiceAccountToken=false for any default SA found for any system NS.

Reply:
But if a system-ns without a default SA exists, won't that error out here with a NotFound error ? Since the handler is off of the NS type and not off of the service-account type, we won't know until we get here.

Reply:
I think we can check if it is a NotFound error and just log this at INFO. We have these cases that fall into this category:
a) a system NS with NO default SA at all- we will just log this at INFO in each sync and not log at error.
b) a system NS which has a default SA but lister failed to list - We will be logging at INFO and hoping next periodic sync should list this when lister cache gets updated.

Reply:
Never mind on this whole thread, I was under false assumption that a default SA might not exist. Looks like it always does.